ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/Update.hs

```diff
@@ -33,17 +36,72 @@ import           Ouroboros.Consensus.Storage.LedgerDB.DbChangelog
 import           Ouroboros.Consensus.Storage.LedgerDB.DbChangelog.Update
 import           Ouroboros.Consensus.Storage.LedgerDB.ReadsKeySets
 import           Ouroboros.Consensus.Util.IOLike
+import qualified Ouroboros.Network.AnchoredSeq as AS
+
+data LedgerDBUpdate = Prune | Extend
+  deriving (Show)
 
 -- | PRECONDITION: The new 'LedgerDB' must be the result of calling either
 -- 'LedgerDB.ledgerDbSwitch' or 'LedgerDB.ledgerDbPushMany' on the current
 -- 'LedgerDB'.
 setCurrent ::
-     MonadSTM m
+     forall blk m.
+     ( MonadSTM m
+     , IsLedger (LedgerState blk)
+     , HasLedgerTables (LedgerState blk)
+     , HasCallStack
+     )
   => StrictTVar m (DbChangelog' blk)
   -> DbChangelog' blk
+  -> LedgerDBUpdate
   -> STM m ()
-setCurrent = writeTVar
+setCurrent v dblog = \case
+  Prune ->
+    updatePruning (changelogLastFlushedState dblog)
+  Extend ->
+    updateExtending (changelogVolatileStates dblog) (changelogDiffs dblog)
+ where
+   updatePruning :: ExtLedgerState blk EmptyMK -> STM m ()
+   updatePruning lf =
+     let
+       f :: (Ord k, Eq v) => SeqDiffMK k v -> SeqDiffMK k v
+       f (SeqDiffMK seqExtended) =
+         SeqDiffMK $ snd $ DS.splitAtSlot (fromWithOrigin 0 $ pointSlot $ getTip lf) seqExtended
+     in
+       modifyTVar v (\extended ->
+                       DbChangelog {
+                          changelogLastFlushedState = lf
+                        , changelogVolatileStates   = changelogVolatileStates extended
+                        , changelogDiffs            =
+                            mapLedgerTables f (changelogDiffs extended)
+                        })
 
+   updateExtending ::
+        AS.AnchoredSeq
+         (WithOrigin SlotNo)
+         (ExtLedgerState blk EmptyMK)
+         (ExtLedgerState blk EmptyMK)
+     -> LedgerTables (ExtLedgerState blk) SeqDiffMK
+     -> STM m ()
+   updateExtending newVolatile extendedDiffs =
+     let
+       f :: (Ord k, Eq v) => SeqDiffMK k v -> SeqDiffMK k v -> SeqDiffMK k v
+       f (SeqDiffMK prunedSeq) (SeqDiffMK extendedSeq) =
+         SeqDiffMK $
+         if DS.minSlot prunedSeq == DS.minSlot extendedSeq
+         then extendedSeq
+         else
+           let (imm, _) = DS.splitAtSlot (fromWithOrigin 0 $ pointSlot $ getTip $ AS.anchor newVolatile) prunedSeq
+               (_, rest) = DS.splitAtFromEnd (AS.length newVolatile) extendedSeq
+           in DS.append imm rest
+     in
+       modifyTVar v (\pruned ->
```

Comment:
Perhaps we should give `v` a more descriptive name. 

Also, I wonder why we call the changelog inside `v` `pruned`. Is to denote that we expect this changelog to be pruned already?